File: chazm-model/src/main/java/io/github/runtimemodels/chazm/id/ClassId.java
package io.github.runtimemodels.chazm.id;

import com.google.inject.assistedinject.Assisted;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;
import java.util.Objects;

import static io.github.runtimemodels.chazm.validation.Checks.checkNotNull;

class ClassId<T> extends AbstractId<T> {

    private static final long serialVersionUID = 5751013993559212419L;
    private final Class<?> id;
    private transient Integer hashCode = null;
    private transient String toString = null;

    @Inject
    ClassId(@NotNull @Assisted final Class<T> type, @NotNull @Assisted final Class<?> id) {
        super(type);
        checkNotNull(id, "id"); //$NON-NLS-1$
        this.id = id;
    }

    @Override
    public boolean equals(final Object object) {
        if (object instanceof ClassId) {
            final ClassId<?> other = (ClassId<?>) object;
            return super.equals(other) && getId().equals(other.getId());
        }
        return false;
    }

    @Override
    public int hashCode() {
        if (hashCode == null) {
            hashCode = Objects.hash(super.hashCode(), getId());
        }
        return hashCode;
    }

    @Override
    public String toString() {
        if (toString == null) {
            toString = super.toString() + ":" + getId();
        }
        return toString;
    }

    public Class<?> getId() {
        return this.id;
    }
}

File: chazm-model/src/main/java/runtimemodels/chazm/model/id/ClassId.kt
package runtimemodels.chazm.model.id

import com.google.inject.assistedinject.Assisted
import java.util.*
import javax.inject.Inject
import javax.validation.constraints.NotNull

@Deprecated("")
internal class ClassId<T> @Inject constructor(
    @Assisted type: Class<T>,
    @param:NotNull @param:Assisted val id: Class<*>
) : AbstractId<T>(type) {
    override fun equals(other: Any?): Boolean {
        if (other is ClassId<*>) {
            return super.equals(other) && id == other.id
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(super.hashCode(), id)

    override fun toString(): String = "${super.toString()}:$id"

    companion object {
        private const val serialVersionUID = 5751013993559212419L
    }
}
